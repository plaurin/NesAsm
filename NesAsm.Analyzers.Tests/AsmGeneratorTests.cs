using ApprovalTests.Reporters;

namespace NesAsm.Analyzers.Tests;

[UseReporter(typeof(VisualStudioReporter))]
public class AsmGeneratorTests
{
    [Fact]
    public void TestExampleScript()
    {
        var generator = new AsmGenerator();
        var content = File.ReadAllText($@"..\..\..\..\NesAsm.Example\HiloWorld.cs");

        GeneratorTestUtilities.TestGenerator(generator, content);
    }

    [Fact]
    public void TestSimpleScript()
    {
        TestScript("SimpleScript.cs");
    }

    [Fact]
    public void TestMultiProcScript()
    {
        TestScript("MultiProcScript.cs");
    }

    [Fact]
    public void TestJumpSubroutineScript()
    {
        TestScript("JumpSubroutineScript.cs");
    }

    [Fact]
    public void TestCharsScript()
    {
        TestScript("CharsScript.cs");
    }

    [Fact]
    public void TestBranchScript()
    {
        TestScript("BranchScript.cs");
    }

    [Fact]
    public void TestParametersScript()
    {
        TestScript("ParametersScript.cs");
    }

    [Fact]
    public void TestSubroutineResultScript()
    {
        TestScript("SubroutineResultScript.cs");
    }

    [Fact]
    public void TestSubroutineResultsScript()
    {
        TestScript("SubroutineResultsScript.cs");
    }

    [Fact]
    public void TestCallingOtherScript()
    {
        TestScript("CallingOtherScript.cs");
    }

    [Fact]
    public void TestDataScript()
    {
        TestScript("DataScript.cs");
    }

    [Fact]
    public void TestAllInstructionsScript()
    {
        TestScript("AllInstructionsScript.cs");
    }

    private static void TestScript(string filename)
    {
        var generator = new AsmGenerator();
        var content = File.ReadAllText($@"..\..\..\TestFiles\{filename}");

        GeneratorTestUtilities.TestGenerator(generator, content);
    }
}
