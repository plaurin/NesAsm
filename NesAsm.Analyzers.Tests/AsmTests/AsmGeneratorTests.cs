using ApprovalTests.Reporters;

namespace NesAsm.Analyzers.Tests.AsmTests;

[UseReporter(typeof(VisualStudioReporter))]
public class AsmGeneratorTests
{
    [Fact]
    public void TestExampleScript()
    {
        var generator = new AsmGenerator();

        GeneratorTestUtilities.TestGenerator(generator, $@"..\..\..\..\NesAsm.Example\HiloWorld\HiloWorld.cs");
    }

    [Fact]
    public void TestSimpleScript()
    {
        TestScript("SimpleScript.cs");
    }

    [Fact]
    public void TestMultiProcScript()
    {
        TestScript("MultiProcScript.cs");
    }

    [Fact]
    public void TestJumpSubroutineScript()
    {
        TestScript("JumpSubroutineScript.cs");
    }

    [Fact]
    public void TestCharsScript()
    {
        TestScript("CharsScript.cs");
    }

    [Fact]
    public void TestBranchScript()
    {
        TestScript("BranchScript.cs");
    }

    [Fact]
    public void TestParametersScript()
    {
        TestScript("ParametersScript.cs");
    }

    [Fact]
    public void TestSubroutineResultScript()
    {
        TestScript("SubroutineResultScript.cs");
    }

    [Fact]
    public void TestSubroutineResultsScript()
    {
        TestScript("SubroutineResultsScript.cs");
    }

    [Fact]
    public void TestCallingOtherScript()
    {
        TestScript("CallingOtherScript.cs", "MultiProcScript.cs");
    }

    [Fact]
    public void TestDataScript()
    {
        TestScript("DataScript.cs");
    }

    [Fact]
    public void TestDataGeneratorScript()
    {
        TestScript("DataGeneratorScript.cs");
    }

    [Fact]
    public void TestAllInstructionsScript()
    {
        TestScript("AllInstructionsScript.cs");
    }

    [Fact]
    public void TestStartupScript()
    {
        TestScript("StartupScript.cs");
    }

    [Fact]
    public void TestIterationStatementsScript()
    {
        TestScript("IterationStatementsScript.cs");
    }

    [Fact]
    public void TestSelectionStatementsScript()
    {
        TestScript("SelectionStatementsScript.cs");
    }

    private static void TestScript(params string[] filenames)
    {
        var generator = new AsmGenerator();

        var sourcePaths = filenames.Select(filename => $@"..\..\..\AsmTests\TestFiles\{filename}").ToArray();

        GeneratorTestUtilities.TestGenerator(generator, sourcePaths);
    }
}
