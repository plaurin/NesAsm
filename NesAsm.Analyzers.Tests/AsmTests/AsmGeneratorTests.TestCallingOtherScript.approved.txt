- Main diagnostics:
- SyntaxTrees Count: 8
- Output diagnostics:
- RunResult.GeneratedTrees Count: 4
- RunResults diagnostics:
- RunResult[0].Generator: Microsoft.CodeAnalysis.IncrementalGeneratorWrapper
- RunResult[0] diagnostics:
- RunResult[0].GeneratedSources Count: 4
- RunResult[0].Exception: 
- RunResult[0].GeneratedSources[0].HintName: Asm.CallingOtherScript.cs
- RunResult[0].GeneratedSources[0].SourceText:
/* Generator Asm code in file Asm.CallingOtherScript.s
; Auto generated code using the NesAsm project
.include "MultiProcScript.s"

.include "../UpFolderScript.s"

.include "SubFolder/SubFolderScript.s"

.scope CallingOtherScript

  .proc main
    jsr MultiProcScript::procB

    jsr myProcC

    jmp MultiProcScript::procC

    rts
  .endproc

  .proc myProcC
    lda MultiProcScript::Data
    lda MultiProcScript::sprite_palettes, x

    jsr UpFolderScript::procU
    jsr SubFolderScript::procS

    rts
  .endproc

.endscope


*/
- RunResult[0].GeneratedSources[1].HintName: Asm.MultiProcScript.cs
- RunResult[0].GeneratedSources[1].SourceText:
/* Generator Asm code in file Asm.MultiProcScript.s
; Auto generated code using the NesAsm project
.scope MultiProcScript

  Data = $FF
.segment "CODE"

sprite_palettes:
  .byte 15
  .byte 32
  .byte 15
  .byte 15
  .byte 15
  .byte 17
  .byte 15
  .byte 15

  .byte 15
  .byte 21
  .byte 15
  .byte 15

  .proc procA
    lda #10

    rts
  .endproc

  .proc procB
    lda #20

    rts
  .endproc

  .proc procC
    lda #30

    rts
  .endproc

.endscope


*/
- RunResult[0].GeneratedSources[2].HintName: Asm.UpFolderScript.cs
- RunResult[0].GeneratedSources[2].SourceText:
/* Generator Asm code in file Asm.UpFolderScript.s
; Auto generated code using the NesAsm project
.scope UpFolderScript

  .proc procU
    ldx #25

    rts
  .endproc

.endscope


*/
- RunResult[0].GeneratedSources[3].HintName: Asm.SubFolderScript.cs
- RunResult[0].GeneratedSources[3].SourceText:
/* Generator Asm code in file Asm.SubFolderScript.s
; Auto generated code using the NesAsm project
.scope SubFolderScript

  .proc procS
    ldx #25

    rts
  .endproc

.endscope


*/
